src/headless/converse-core.js

```diff
@@ -71,27 +71,42 @@ const _converse = {
     'templates': {},
     'promises': {},
 
-    get connection () {
-        /* Creates a new Strophe.Connection instance if we don't already have one.
-         */
-        if (!this._connection) {
-            if (!this.bosh_service_url && ! this.websocket_url) {
-                throw new Error("connection: you must supply a value for either the bosh_service_url or websocket_url or both.");
+    onDomainDiscovered (ev) {
+        const xrd = ev.target.responseXML.documentElement;
+        if (xrd.nodeName != "XRD" || xrd.namespaceURI != "http://docs.oasis-open.org/ns/xri/xrd-1.0") {
+            _converse.log("Could not discover XEP-0156 connection methods", Strophe.LogLevel.WARN);
+            return initConnection();
+        }
+        const bosh_methods = [];
+        const websocket_methods = [];
+        for (let i = 0; i < xrd.children.length; i++) {
+            const child = xrd.children[i];
+            if (child.nodeName != "Link" || child.namespaceURI != "http://docs.oasis-open.org/ns/xri/xrd-1.0") {
+                continue;
             }
-            if (('WebSocket' in window || 'MozWebSocket' in window) && this.websocket_url) {
-                this._connection = new Strophe.Connection(this.websocket_url, this.connection_options);
-            } else if (this.bosh_service_url) {
-                this._connection = new Strophe.Connection(
-                    this.bosh_service_url,
-                    _.assignIn(this.connection_options, {'keepalive': this.keepalive})
-                );
-            } else {
-                throw new Error("connection: this browser does not support websockets and bosh_service_url wasn't specified.");
+            const rel = child.getAttributeNS(null, "rel");
+            const href = child.getAttributeNS(null, "href");
+            if (rel == "urn:xmpp:alt-connections:xbosh") {
+                bosh_methods.push(href);
+            } else if (rel == "urn:xmpp:alt-connections:websocket") {
+                websocket_methods.push(href);
             }
-            _converse.emit('connectionInitialized');
         }
-        return this._connection;
-    }
+        if (websocket_methods.length > 0) {
```

Comment:
Note that these two checks were hacks because I didn’t want to implement RR selection of the endpoint just yet.